### PR TITLE
Plan `LATERAL` subqueries

### PR DESCRIPTION
**Pull request by aalexandrov**
## Which issue does this PR close?

Closes the planning part of #10048. The query from the issue will now fail in the optimization phase similar to #6140.

## Rationale for this change

Adds an unsupported fragment of the SQL syntax.

## What changes are included in this PR?

Extends the planner with support for `LATERAL` subqueries. See the tests added in `sql_integration.rs` for details.

## Are these changes tested?

Yes, I added tests in `sql_integration.rs`.

## Are there any user-facing changes?

If there are user-facing changes then we may require documentation to be updated before approving the PR.

If we document the supported SQL fragment in the user-facing docs as BNF-style diagrams we might want to adapt those.

I don't think so.

### CODE REVIEW

**Comment by ozankabak** [comment]
Thank you! IMO, missing lateral and correlated subquery support is one of the crucial things to fix in DF. Any work in this area is much appreciated.

File: datafusion/sql/src/relation/mod.rs [hunk 1]
```diff
@@ -143,4 +147,45 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
...
+    ) -> Result<LogicalPlan> {
+        // At this point for a syntacitally valid query the outer_from_schema is
+        // guaranteed to be set, so the `.unwrap()` call will never panic. This
+        // is the case because we only call this method for lateral table
+        // factors, and those can never be the first factor in a FROM list. This
+        // means we arrived here through the `for` loop in `plan_from_tables` or
+        // the `for` loop in `plan_table_with_joins`.
+        let old_from_schema = planner_context.set_outer_from_schema(None).unwrap();
+        let new_query_schema = match planner_context.outer_query_schema() {
+            Some(lhs) => Some(Arc::new(lhs.join(&old_from_schema)?)),
```
Comment (aalexandrov) [inline datafusion/sql/src/relation/mod.rs:164]: We probably want to replace the `lhs.join(<rhs>)` call here with aversion of a `join` where duplicate columns on the right side override their `lhs` counterpart.

See #11464 for details.

File: datafusion/sql/src/planner.rs [hunk 2]
```diff
@@ -151,6 +155,29 @@ impl PlannerContext {
...
+        mut schema: Option<DFSchemaRef>,
+    ) -> Option<DFSchemaRef> {
+        std::mem::swap(&mut self.outer_from_schema, &mut schema);
+        schema
+    }
+
+    /// extends the FROM schema, returning the existing one, if any
+    pub fn extend_outer_from_schema(&mut self, schema: &DFSchemaRef) -> Result<()> {
+        self.outer_from_schema = match self.outer_from_schema.as_ref() {
+            Some(from_schema) => Some(Arc::new(from_schema.join(schema)?)),
```
Comment (aalexandrov) [inline datafusion/sql/src/planner.rs:175]: See [my comment below](https://github.com/apache/datafusion/pull/11456/files#r1677146927).

**Comment by alamb** [comment]
Thank you @aalexandrov -- I will review this PR carefully, but likely will not have a chance until tomorrow or Tuesday

File: datafusion/sql/src/relation/mod.rs (hunk 1 above)
Comment (aalexandrov) [reply datafusion/sql/src/relation/mod.rs]: I added a second commit to the PR that fixes the issue mentioned above (see the commit message for details).

File: datafusion/sql/src/relation/mod.rs [hunk 3]
```diff
@@ -143,4 +147,49 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
...
+        subquery: TableFactor,
+        planner_context: &mut PlannerContext,
+    ) -> Result<LogicalPlan> {
+        // At this point for a syntacitally valid query the outer_from_schema is
+        // guaranteed to be set, so the `.unwrap()` call will never panic. This
+        // is the case because we only call this method for lateral table
+        // factors, and those can never be the first factor in a FROM list. This
+        // means we arrived here through the `for` loop in `plan_from_tables` or
+        // the `for` loop in `plan_table_with_joins`.
+        let old_from_schema = planner_context.set_outer_from_schema(None).unwrap();
```
Comment (jonahgao) [inline datafusion/sql/src/relation/mod.rs:162]: I tested the following query, and it panicked. It's valid in PostgreSQL.
```sh
DataFusion CLI v40.0.0

> select * from lateral (select 1) t;
thread 'main' panicked at datafusion/datafusion/sql/src/relation/mod.rs:162:75:
called `Option::unwrap()` on a `None` value
```

**Comment by alamb** [comment]
This PR is very much on my list to review, but I fear I may not get to it today (I have a bunch of PRs waiting upstream in arrow-rs I need to attend to)

File: datafusion/sql/src/relation/mod.rs (hunk 3 above)
Comment (aalexandrov) [reply datafusion/sql/src/relation/mod.rs]: Somehow I was thinking that `LATERAL` is not valid / accepted by the parser for the first element in a `SELECT` list. I fixed the issue and added a regression test in `sql_integration.rs`.

**Comment by alamb** [comment]
Still on my list... .

I am not sure if @jackwener  or @mingmwang  might have time to review this (or if they are familiar with this particular rewrite)

**Review by alamb — COMMENTED** [review]
Thanks @aalexandrov  -- sorry for not getting a chance to review this PR earlier (I have been quite busy with other stuff and JOINs are pretty low on my priority list unfortunately)

The basic code in this PR looks good to me-- but I don't know enough about the semantics of lateral joins to know if it is 

Can you please add some tests to sqllogictest that shows that this code gets the same code in postgres and in DataFusion

Here are the instructions: https://github.com/apache/datafusion/tree/main/datafusion/sqllogictest

The instructions for postgres compatibility tests are here
https://github.com/apache/datafusion/tree/main/datafusion/sqllogictest#running-tests-postgres-compatibility

Ideally you should be able to extend one of the existing test files in https://github.com/apache/datafusion/tree/main/datafusion/sqllogictest/test_files

**Comment by aalexandrov** [comment]
> Can you please add some tests to sqllogictest that shows that this code gets the same code in postgres and in DataFusion.

This PR only focuses on extending `datafusion/sql/src/planner.rs` so we can get logical plans for the `LATERAL` syntax.

I was hoping to merge that first before taking a stab at the required extensions in `datafusion/sql/src/planner.rs` that would allow us to run those queries. Once we have those I will be able to add some SLT tests and fix any bugs in both files.

**Comment by alamb** [comment]
> I was hoping to merge that first before taking a stab at the required extensions in datafusion/sql/src/planner.rs that would allow us to run those queries. Once we have those I will be able to add some SLT tests and fix any newly uncovered bugs in both files.

Implementing in phases makes sense to me and is consistent with what we have done in other places

However, I still think  we should add basic .slt test coverage (even if they simply error our with "unsupported subquery") is to document, with code, what the expected behavior is if someone runs this kind of query. 

I think a runtime "not yet implemented" error is fine, but going into infinite loops or internal errors is probably less good.

**Comment by aalexandrov** [comment]
I added some tests to check the behavior of the SQL planner and assert the current runtime behavior (unimplemented because physical planning doesn't know what to do with correlated column references) to `joins.slt`.

**Review by alamb — APPROVED** [review]
Thank you @aalexandrov  - this code and tests look good to me.

I again apologize for not having a chance to review this PR sooner. 

All in all, I think this is a good step forward.

I took the liberty of merging this PR up from main to resolve a conflict and fixed some expected tests

cc @jonahgao  in case you are interested in reviewing the SQL planning aspects

**Comment by alamb** [comment]
🚀

**Comment by alamb** [comment]
Thanks again for bearing with us @aalexandrov  -- this is a great step

**Comment by aalexandrov** [comment]
Thanks, @alamb. Can you point me to the unnesting rules that already exist? I might try to extend them to cover some of the more basic queries first.

**Comment by alamb** [comment]
> Thanks, @alamb. Can you point me to the unnesting rules that already exist? I might try to extend them to cover some of the more basic queries first.

The optimizer rules are here: https://docs.rs/datafusion/latest/datafusion/optimizer/optimizer/index.html

I think in particular two that are relevant are:
* https://docs.rs/datafusion/latest/datafusion/optimizer/scalar_subquery_to_join/index.html
* https://docs.rs/datafusion/latest/datafusion/optimizer/decorrelate_predicate_subquery/index.html

**Pull request merged**